core/src/test/java/org/apache/iceberg/TestRowLineageAssignment.java

```diff
@@ -0,0 +1,672 @@
+    InputFile mergedManifestList =
+        upgradeTable.io().newInputFile(upgradeTable.currentSnapshot().manifestListLocation());
+    checkManifestListAssignment(mergedManifestList, 0L);
+    List<ManifestFile> mergedManifests =
+        upgradeTable.currentSnapshot().dataManifests(upgradeTable.io());
+    checkDataFileAssignment(
+        upgradeTable,
+        mergedManifests.get(0),
+        0L,
+        FILE_C.recordCount(),
+        FILE_C.recordCount() + FILE_B.recordCount());
+  }
+
+  private ManifestContent content(int ordinal) {
+    return ManifestContent.values()[ordinal];
+  }
+
+  private void checkManifestListAssignment(InputFile in, Long... firstRowIds) {
+    try (CloseableIterable<Record> reader =
+        InternalData.read(FileFormat.AVRO, in)
+            .project(ManifestFile.schema().select("first_row_id", "content"))
+            .build()) {
+
+      // all row IDs must be assigned at write time
+      int i = 0;
+      for (Record manifest : reader) {
+        if (content((Integer) manifest.getField("content")) != ManifestContent.DATA) {
+          assertThat(manifest.getField("first_row_id"))
+              .as("Row ID for delete manifest (%s) should be null", i)
+              .isNull();
+        } else if (i < firstRowIds.length) {
+          assertThat(manifest.getField("first_row_id"))
+              .as("Row ID for data manifest (%s) should match", i)
+              .isEqualTo(firstRowIds[i]);
+        } else {
+          fail("No expected first row ID for manifest: %s=%s", i, manifest);
+        }
+
+        i += 1;
+      }
+
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    // also check that the values are read correctly
+    int i = 0;
+    for (ManifestFile manifest : ManifestLists.read(in)) {
+      if (manifest.content() != ManifestContent.DATA) {
+        assertThat(manifest.firstRowId()).isNull();
+      } else if (i < firstRowIds.length) {
+        assertThat(manifest.firstRowId()).isEqualTo(firstRowIds[i]);
+      } else {
+        fail("No expected first row ID for manifest: " + manifest);
+      }
+
+      i += 1;
+    }
+  }
+
+  private void checkDataFileAssignment(Table table, ManifestFile manifest, Long... firstRowIds) {
+    // all row IDs must be assigned at write time
+    int i = 0;
+    try (ManifestReader<DataFile> reader =
+        ManifestFiles.read(manifest, table.io(), table.specs())) {
+
+      for (DataFile file : reader) {
+        assertThat(file.content()).isEqualTo(FileContent.DATA);
+        if (i < firstRowIds.length) {
+          assertThat(file.firstRowId())
+              .as("Row ID for data file (%s) should match", i)
+              .isEqualTo(firstRowIds[i]);
+        } else {
+          fail("No expected first row ID for file: " + manifest);
+        }
+
+        i += 1;
+      }
+
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
```

Comment:
Not sure if we need to check this, but the other cases I can imagine testing are cherry-pick/merge of a preV3 branches as the first operation after upgrade.

IE
main1 -> main2
branch1

Upgrade
Cherrypick/Merge Branch1
main1-main2-branch1
This should do a full assignment I believe

Reply:
I think we also have coverage of "delete" after upgrade with Overwrite. But it may be nice to add those tests as well.

File 1 and 2 Exists
Upgrade
Delete File1

File2 should be assigned

Reply:
Adding `testDeleteAssignmentAfterUpgrade` with the case you suggested.

For the cherry-pick case, I think that we should cover it in a follow up if that's alright. That's different enough functionality and I'm hesitant to make this any larger.

Reply:
That all makes sense, feel free to also just make some new issues and tag them as good first issues. I think this is good work for a new contributor who can follow the old patterns and write new tests. I don't mean to put all the work on you